In [1]:
import pandas as pd
import numpy as np
df= pd.read_csv('datasets/credit_scoring_eng.csv')
df.head()

children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose  
0    purchase of the house  
1             car purchase  
2    purchase of the house  
3  supplementary education  
4        to have a wedding

(Updated) The data give information about people who come from different groups. People are categorized according to, number of children that they have, number of their workdays,  age, education gender, family status, total income, have a debt problem, and the reason why they have a loan. Our task is to find whether there is a correlation among these people: In a simple form, for example, men or women work more or which group earns overall more.  In the data set, the categories are shown in columns 

Step 2. Data preprocessing

In [2]:
df.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

I used isna  and sum method to check if there is a missing value.

In [3]:
df.isna().mean().round(4) * 100

children             0.0
days_employed       10.1
dob_years            0.0
education            0.0
education_id         0.0
family_status        0.0
family_status_id     0.0
gender               0.0
income_type          0.0
debt                 0.0
total_income        10.1
purpose              0.0
dtype: float64

I wanted to know the weather percentages of missing value  is significant . Then We see that yes , it is significant and we should not delete them.

Data type replacement

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [5]:
#(Updated)
df['days_employed']=np.int64(df['days_employed'])
df['total_income']=np.int64(df['total_income'])

                                                    Processing duplicates

In [6]:
for col in df:
    print('                            Unique value of for')
    print("|_______________________________________________________________________________|")

    
    print(df[col].unique())
    print('________________________________________________________________________________')

                            Unique value of for
|_______________________________________________________________________________|
[ 1  0  3  2 -1  4 20  5]
________________________________________________________________________________
                            Unique value of for
|_______________________________________________________________________________|
[ -8437  -4024  -5623 ... 362161 373995 343937]
________________________________________________________________________________
                            Unique value of for
|_______________________________________________________________________________|
[42 36 33 32 53 27 43 50 35 41 40 65 54 56 26 48 24 21 57 67 28 63 62 47
 34 68 25 31 30 20 49 37 45 61 64 44 52 46 23 38 39 51  0 59 29 60 55 58
 71 22 73 66 69 19 72 70 74 75]
________________________________________________________________________________
                            Unique value of for
|__________________________________________________________________

I used this code to  see all unique values, beacuse I can know check if all data were correctly filled .


In [7]:
df['days_employed']=df['days_employed'].apply(abs)
df['total_income']=df['total_income'].apply(abs)
df['children']=df['children'].apply(abs)

I convert all value into positive because the negative value is these columns (because of the information they have) is not possible

In [8]:
df.loc[df['children'] == 20, 'children'] =2

Having 20 children does not seem to be  real, but may be in Carone time, it  is possible.

In [9]:

df['days_employed'] = df['days_employed'].fillna(df.groupby('income_type')['days_employed'].transform('mean'))
df['total_income'] = df['total_income'].fillna(df.groupby('income_type')['total_income'].transform('mean'))


(Updated) I replace the missing value with the mean value of columsn group by the income level. Beacuse In this case we can get more realistic mean number.(Updated)

In [10]:
# reviewers code
df['days_employed'].agg(['min','max'])

min                     24
max    9223372036854775808
Name: days_employed, dtype: uint64

In [11]:
df.loc[df['dob_years'] == 0, 'dob_years'] =int(df['dob_years'].mean())

There is a zero aga, but of course is not possibe , therefore I replace it with mean value of all ages.

Delete duplicate data"

In [12]:
df['education'] = df['education'].str.lower()
df['purpose'] = df['purpose'].str.lower()

Then we can see they are different information or they are just the same thing but Capital letters are different. 

In [13]:
df.sort_values("total_income", inplace = True)
df.duplicated().sum()
df = df.drop_duplicates().reset_index(drop=True)

In [14]:
df.duplicated().sum()

0

I dropped all duplicate values because duplicate values could lead to misleading results.

In [15]:
for col in df:
    print('                            Unique value of for')
    print("|_______________________________________________________________________________|")

    
    print(df[col].unique())
    print('________________________________________________________________________________')

                            Unique value of for
|_______________________________________________________________________________|
[0 1 2 3 5 4]
________________________________________________________________________________
                            Unique value of for
|_______________________________________________________________________________|
[             359219              369708                3642 ...
               10207                5734 9223372036854775808]
________________________________________________________________________________
                            Unique value of for
|_______________________________________________________________________________|
[57 37 52 68 61 71 60 43 55 45 46 59 67 28 49 64 50 69 70 27 44 62 54 34
 26 24 72 63 35 53 32 21 39 41 65 48 58 29 56 33 47 66 38 40 42 51 74 36
 31 23 30 25 22 19 20 73 75]
________________________________________________________________________________
                            Unique value of for
|__

In [16]:
df.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

Last time to check data for missing value and duplicate value.

In [17]:
df.shape

(21454, 12)

(Updated) Okay, then data is cleaned from unnecessary data, by deleting duplicated values and filling missing values with mean values. Also, we convert some data formats into the correct ones. The data is ready to work on it. The next task will be caterofiring data in proper ways to find correction amounts in these groups and see if this correlation is significant. I type shape to see change in our data . The number of row is less now.

                                                                      Categorizing Data

In [18]:
df

children        days_employed  dob_years            education  \
0             0               359219         57  secondary education   
1             0               369708         37  secondary education   
2             1                 3642         52  secondary education   
3             0               359726         68  secondary education   
4             0               346602         61  secondary education   
...         ...                  ...        ...                  ...   
21449         0  9223372036854775808         68  secondary education   
21450         2  9223372036854775808         50         some college   
21451         1  9223372036854775808         39  secondary education   
21452         2  9223372036854775808         30  secondary education   
21453         2  9223372036854775808         31  secondary education   

       education_id      family_status  family_status_id gender income_type  \
0                 1            married                 0      F     retiree   
1                 1  civil partnership                 1      M     retiree   
2                 1            married                 0      M    employee   
3                 1  civil partnership                 1      M     retiree   
4                 1            married                 0      F     retiree   
...             ...                ...               ...    ...         ...   
21449             1            married                 0      F     retiree   
21450             2            married                 0      F    business   
21451             1            married                 0      F    employee   
21452             1  civil partnership                 1      F    employee   
21453             1            married                 0      F    employee   

       debt         total_income                      purpose  
0         1                 3306                     property  
1         0                 3392          going to university  
2         0                 3418                 car purchase  
3         0                 3471             having a wedding  
4         0                 3503                     property  
...     ...                  ...                          ...  
21449     0  9223372036854775808  buy residential real estate  
21450     0  9223372036854775808     purchase of my own house  
21451     0  9223372036854775808   buy commercial real estate  
21452     0  9223372036854775808            to have a wedding  
21453     0  9223372036854775808     buying a second-hand car  

[21454 rows x 12 columns]

                                                              Categorizing Data

In [19]:
def age_group(dob_years):
    
    """
    The function returns the age group according to the age value, using the following rules:
    —'children' for age <= 18
    —'adult' for 19 <= age <= 64
    —'retired' for all other cases
    """
    if dob_years <= 18:
        return 'children' 
    if dob_years <= 64:
        return 'adult'
    return 'retired' 

df['dob_years'] = df['dob_years'].apply(age_group)

I categorized ages into 3 groups, We can see that there is no child in our data. categorizing is beneficial because it makes it possible to focus on specific groups and find the correlation. For example the next command in line 19 and 20 
:

In [20]:
df['total_income'] = pd.to_numeric(df['total_income'], errors='coerce')
 
def income_level(total_income):
    """
Low-income - under 10,000
Lower-middle income - 10,001 - 20,000
Upper-middle income - 20,001 - 30,000
High income > 30,001
    """
 
    if total_income <= 10000:
        return 'low_income'
    if total_income <= 20000:
        return 'mid_income'
    if total_income <= 30000:
        return 'high_income'
    return 'very_high_income'
 
df['income_level'] = df['total_income'].apply(income_level)


In [21]:
df['income_level'].value_counts()

very_high_income    8022
mid_income          6443
high_income         6062
low_income           927
Name: income_level, dtype: int64

we can also categorized people according to their income level.

In [22]:
def purpose_loan_out(purpose):
 
    if purpose == 'construction of own property' or purpose=='building a property' or purpose=='buying property for renting out'or purpose=='property' :
        return 'property'
    if purpose=='going to university' or purpose=='to become educated' or purpose=='getting an education' or purpose=='profile education' or purpose=='getting higher education' or purpose=='to get a supplementary education' or purpose=='education'  :
        return 'education'
    if purpose == 'cars' or purpose=='purchase of a car' or purpose=='buying my own car' or purpose=='buying a second-hand car'or purpose=='second-hand car purchase'or purpose=='to own a car'  or purpose=='to buy a car' or purpose=='car purchase':
        return 'car purchase'
    if purpose=='having a wedding' or purpose=='to have a wedding' or purpose=='wedding ceremony' :
        return 'wedding ceremony'
    if purpose=='transactions with my real estate' or purpose=='housing' or purpose=='purchase of the house for my family' or purpose=='purchase of my own house' or purpose=='buy real estate' or purpose=='purchase of the house' or purpose=='building a real estate' :
        return 'purchase of my own house'
    if purpose=='transactions with commercial real estate' or purpose=='buy commercial real estate' or purpose=='real estate transactions':
        return 'purchase house for rent'
    if purpose=='housing renovation':
        return 'housing renovation'
            
            
df['purpose_loan_out'] = df['purpose'].apply(purpose_loan_out)    


In [23]:
df['purpose_loan_out'].value_counts()

purchase of my own house    4422
car purchase                3812
education                   3101
property                    2538
wedding ceremony            2324
purchase house for rent     1986
housing renovation           607
Name: purpose_loan_out, dtype: int64

There is a lot of the same information in data, by categorizing them in 7 groups, it will be easier to compare.

In [24]:
df.groupby('dob_years')['total_income'].sum()

dob_years
adult      1.868655e+22
retired    7.101996e+20
Name: total_income, dtype: float64

In [25]:
df.groupby('dob_years').sum()

children  days_employed  education_id  family_status_id  debt  \
dob_years                                                                  
adult         10268   1.868655e+22         16695             19925  1692   
retired          42   7.101996e+20           835               969    49   

           total_income  
dob_years                
adult      1.868655e+22  
retired    7.101996e+20

Here we see that adult's income is larger than retired's income .

We can also use these categorized data to find out, which group we have to focus on more. For example, we see that adults have more debt, we can go them some idea about debt.

(Updated) We categorized data according to age group, income level, and purpose for applying for a loan. Then we can compare the data easily because we can focus on the specific data groups.  We have now a compacted data set. The next task will be to dive more into data to find relations between different groups.

                                                                          Step 3. Answer these questions:

In [26]:
df.groupby('children')['debt'].mean().round(3)*100

children
0    7.5
1    9.2
2    9.5
3    8.2
4    9.8
5    0.0
Name: debt, dtype: float64

Is there a relation between having kids and repaying a loan on time

(Updated) We can see that there 9.8 percent of people who have 4 children have also a problem paying back the loan. But people who just have one more child have almost no problem with debt. Therefore, it is clear that there is no correlation between the number of children and haveing problems with paying back the debt.

In [27]:
df.groupby('family_status')['debt'].mean().round(3)*100

family_status
civil partnership    9.3
divorced             7.1
married              7.5
unmarried            9.8
widow / widower      6.6
Name: debt, dtype: float64


Is there a relation between marital status and repaying a loan on time?

civil partnership and  unmarried group have the the higest amount of debt.

In [28]:
df.groupby('income_level')['debt'].mean().round(3)*100

income_level
high_income         8.7
low_income          6.3
mid_income          8.5
very_high_income    7.6
Name: debt, dtype: float64

Is there a relation between marital status and repaying a loan on time?

#There is no increasing or decreasing trend as people becoming richer. 
Updated. We see that people who have high income , have also most problem paying back the debt in time . This number is lest in low income goup . people in very high income group have less problem with debt in comparision with high income.

In [29]:
df.groupby('purpose_loan_out')['debt'].mean().round(3)*100

purpose_loan_out
car purchase                9.5
education                   9.4
housing renovation          5.8
property                    7.5
purchase house for rent     7.8
purchase of my own house    7.1
wedding ceremony            8.0
Name: debt, dtype: float64

Is there a relation between income level and repaying a loan on time?

  (Updated)  The data shows that most of the people who get a loan and cannot pay in time back wanted to buy a car or to get an education. The next group which has a problem with a loan is who gets a loan for the wedding ceremony.  The least a number of people who paid the debt, not in time, get the loan for housing renovation. I would say that people who have relatively less problem with paying their loan in time, mainly get this credit for house or property related reasons.

Step 4. General conclusion

The data is about to research people who have different backgrounds. After cleaning the data unnecessary information these are 21454 rows and 12 columns. So the data set has a significant amount of information to come to a conclusion. Then data are categorized into different groups such as income level, the purpose of the loan, and age group. They compared these categories. In other to find a relation among the groups I grouped the columns. Then I see that there a correlation between income level, the purpose of the loan, family status, and having a problem with paying the debt in time. For me, it is interesting that most people who get the debt to buy a car have a problem with paying their loan in time. And also higher income group have the same problem.  